### PR TITLE
Periodic background reporter for "Server" metrics.

### PR DESCRIPTION
**Pull request by apg**
First of all, I'm not typically a Rubyist, so all comments are welcome in order to get this into tip top shape. Secondly, here is my attempt at getting something landed that addresses my question in #11, in which @jeremy suggested he'd entertain a PR.

I've tried to keep the spirit of the original functionality, but have pulled out counters and gauges that were previously reported as `Rack.Server` and made them be reported periodically. This involves more setup, but is also, arguably, a better, more accurate approach. The per-request data is still being generated as timings emitted per request, like before, provided you set that up as well (the original directions suffice). This *doesn't* do aggregation of the per request metrics, which may, or may not, be the right thing to do here...

### CODE REVIEW

File: lib/trashed/railclock.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+    initializer 'trashed.periodic' do |app|
+      require 'statsd'
+
+      app.config.trashed_periodic.logger ||= Rails.logger
+
+      Periodic.new config.trashed_periodic
+    end
+  end
+end
```
Comment (jeremy) [inline lib/trashed/railclock.rb]: Railties are meant to handle multiple initializers and setup. We're good to merge this into the main Railtie.

File: lib/trashed/reporter.rb [hunk 2]
```diff
@@ -23,39 +23,71 @@ def report(env)
     end
 
     def report_logger(env)
-      timings = env[Trashed::Rack::TIMINGS]
+      raise "must implement interface"
+    end
+
+    def report_statsd(env)
+      raise "must implement interface"
```
Comment (jeremy) [inline lib/trashed/reporter.rb]: `raise NotImplementedError`

File: test/reporter_test.rb [hunk 3]
```diff
@@ -4,15 +4,32 @@
 require 'stringio'
 
 class ReporterTest < Minitest::Test
+
```
Comment (jeremy) [inline test/reporter_test.rb]: ✂️ prefer no extra space after class declaration

File: test/lib/trashed/test_helper.rb [hunk 4]
```diff
@@ -1,2 +1,11 @@
...
 require 'minitest/autorun'
+
+class Statsd
+  def initialize(batcher)
+    @batcher = batcher
+  end
+  def batch
+    yield @batcher
+  end
+end
```
Comment (jeremy) [inline test/lib/trashed/test_helper.rb]: Let's nest this within `PeriodicReporterTest` where it's used.

File: lib/trashed/railclock.rb (hunk 1 above)
Comment (apg) [reply lib/trashed/railclock.rb]: The only issue there is that I want some ability to only run the Periodic reporter, or the PerRequest reporter (or possibly both). Do you have a suggestion for how to manage that?

File: lib/trashed/reporter.rb [hunk 5]
```diff
@@ -74,8 +74,10 @@ def report_logger(env)
     def report_statsd(env)
       method = @statsd.respond_to?(:easy) ? :easy : :batch
       @statsd.send(method) do |statsd|
-        send_to_statsd statsd, :timing, @timing_sample_rate, env[Trashed::Rack::TIMINGS], :'Rack.Request', @timing_dimensions.call(env)
-        send_to_statsd statsd, :timing, @gauge_sample_rate,  env[Trashed::Rack::GAUGES],  :'Rack.Server',  @gauge_dimensions.call(env)
+        # We emit the counters as timings to take advantage of statsd
+        # aggregations, min, max, etc.
+        send_to_statsd statsd, :timing, @counter_sample_rate, env[Trashed::COUNTERS], :'Rack.Request', @counter_dimensions.call(env)
+        send_to_statsd statsd, :timing, @gauge_sample_rate, env[Trashed::GAUGES], :'Rack.Server', @gauge_dimensions.call(env)
```
[the remark this answers is not in the document]
Comment (jeremy) [reply lib/trashed/reporter.rb]: Ping @apg

[the remark this answers is not in the document]
Comment (apg) [reply lib/trashed/reporter.rb]: Bah! Github notifications.

Linking out to that probably makes sense. I'll update.

[the remark this answers is not in the document]
Comment (jeremy) [reply lib/trashed/reporter.rb]: Let's revert the `timings` → `counters` mass name change as well. The Ruby naming directly reflects the StatsD metric types we're using.

[the remark this answers is not in the document]
Comment (apg) [reply lib/trashed/reporter.rb]: Only the per request metrics use `:timing` statsd types though. The PeriodicReporter used `:count` and `:gauge`. I'd argue that the use of `counters` and `gauges` provides semantic meaning to the values that are being collected, even if they *are* reported to statsd in a different manner (that seems irrelevant to me, really).

**Comment by stereobooster** [comment]
Also periodic reporter can track number of threads

**Comment by stereobooster** [comment]
@jeremy anything else required to finish this PR?

**Comment by jeremy** [comment]
@stereobooster @apg https://github.com/basecamp/trashed/pull/12#discussion_r137167160

**Comment by jeremy** [comment]
@apg Are you interested in pursuing this, given that you've forked (#18)? (I'm assuming that you parked it and decided to fork at the naming change.)

**Comment by apg** [comment]
@jeremy I'm not likely to pursue this in trashed, no. It seems like the goals of trashed are fairly different than what we're trying to do in [barnes](https://github.com/heroku/barnes) so our new path makes more sense at this stage in the game.

That being said, @stereobooster, if @jeremy is still willing to accept, please feel free to push on with it.

**Comment by jeremy** [comment]
Cool. Best of luck, @apg! 🚢

**Pull request closed without merging**